File: frontend/src/lib/firmware.ts
import { api, readErrorMessage } from "@/lib/api";

export interface FirmwareStatus {
  current_version: string | null;
  available_version: string | null;
  channel: string | null;
  checked_at: string | null;
  routerboard_current: string | null;
  routerboard_available: string | null;
  needs_upgrade: boolean;
}

export interface FleetFirmwareSummary {
  total: number;
  updates_available: number;
  checked_ever: number;
  never_checked: number;
}

export interface FirmwareUpgradeStatus {
  last_triggered_at: string | null;
  last_status: "pending" | "succeeded" | "failed" | null;
  last_error: string | null;
  last_from_version: string | null;
  last_to_version: string | null;
}

export interface FirmwareUpgradeResult {
  triggered: boolean;
  will_reboot: boolean;
  from_version: string | null;
  to_version: string | null;
  message: string;
}

export interface AutoUpgradePolicy {
  enabled: boolean;
  window_start_hour: number | null;
  window_end_hour: number | null;
}

export async function getFleetFirmwareSummary(): Promise<FleetFirmwareSummary> {
  return api.get("firmware/summary").json<FleetFirmwareSummary>();
}

export async function getDeviceFirmware(deviceId: string): Promise<FirmwareStatus> {
  return api.get(`devices/${deviceId}/firmware`).json<FirmwareStatus>();
}

export async function checkDeviceFirmware(deviceId: string): Promise<FirmwareStatus> {
  try {
    return await api
      .post(`devices/${deviceId}/firmware/check`, { timeout: 60_000 })
      .json<FirmwareStatus>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export type FirmwareUpgradeTarget = "routeros" | "routerboard" | "both";

export async function triggerFirmwareUpgrade(
  deviceId: string,
  opts: { target: FirmwareUpgradeTarget },
): Promise<FirmwareUpgradeResult> {
  try {
    return await api
      .post(`devices/${deviceId}/firmware/upgrade`, {
        json: { target: opts.target },
        timeout: 60_000,
      })
      .json<FirmwareUpgradeResult>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function getFirmwareUpgradeStatus(
  deviceId: string,
): Promise<FirmwareUpgradeStatus> {
  return api
    .get(`devices/${deviceId}/firmware/upgrade-status`)
    .json<FirmwareUpgradeStatus>();
}

export async function getAutoUpgradePolicy(deviceId: string): Promise<AutoUpgradePolicy> {
  return api.get(`devices/${deviceId}/firmware/policy`).json<AutoUpgradePolicy>();
}

export async function setAutoUpgradePolicy(
  deviceId: string,
  policy: AutoUpgradePolicy,
): Promise<AutoUpgradePolicy> {
  try {
    return await api
      .put(`devices/${deviceId}/firmware/policy`, { json: policy })
      .json<AutoUpgradePolicy>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

File: backend/app/schemas/firmware.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class FirmwareStatusPublic(BaseModel):
    current_version: str | None
    available_version: str | None
    channel: str | None
    checked_at: datetime | None
    routerboard_current: str | None
    routerboard_available: str | None
    needs_upgrade: bool


class FleetFirmwareSummary(BaseModel):
    total: int
    updates_available: int
    checked_ever: int
    never_checked: int


class FirmwareUpgradeRequest(BaseModel):
    """Targets are independent so the UI can offer separate buttons:

    - `routeros` (default): upgrade the RouterOS package only.
    - `routerboard`: upgrade only the RouterBOARD bootloader. Useful when
      RouterOS is already on the target version and only the bootloader
      lags behind.
    - `both`: upgrade RouterOS first, then the bootloader (two reboots).

    The legacy `include_routerboard=true` request is still accepted and
    treated as `target="both"`.
    """

    target: Literal["routeros", "routerboard", "both"] = "routeros"
    include_routerboard: bool = False

    @model_validator(mode="after")
    def _coerce_legacy_flag(self) -> "FirmwareUpgradeRequest":
        if self.include_routerboard and self.target == "routeros":
            # Caller used the old single-flag API: promote to "both".
            object.__setattr__(self, "target", "both")
        return self


class FirmwareUpgradeResult(BaseModel):
    triggered: bool
    will_reboot: bool
    from_version: str | None
    to_version: str | None
    message: str


class FirmwareUpgradeStatus(BaseModel):
    last_triggered_at: datetime | None
    last_status: str | None         # "pending" | "succeeded" | "failed"
    last_error: str | None
    last_from_version: str | None
    last_to_version: str | None


class AutoUpgradePolicyPublic(BaseModel):
    enabled: bool
    window_start_hour: int | None
    window_end_hour: int | None


class AutoUpgradePolicyUpdate(BaseModel):
    enabled: bool
    window_start_hour: int | None = Field(default=None, ge=0, le=23)
    window_end_hour: int | None = Field(default=None, ge=0, le=23)

    @model_validator(mode="after")
    def _window_pair(self) -> "AutoUpgradePolicyUpdate":
        # Both or neither — a half-set window is almost certainly user error.
        if (self.window_start_hour is None) ^ (self.window_end_hour is None):
            raise ValueError("window_start_hour and window_end_hour must be set together")
        return self
